homeassistant/components/siren/__init__.py

```diff
@@ -0,0 +1,222 @@
+"""Component to interface with various sirens/chimes."""
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    SERVICE_TOGGLE,
+    SERVICE_TURN_OFF,
+    SERVICE_TURN_ON,
+    STATE_ON,
+)
+from homeassistant.core import ServiceCall
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.config_validation import (  # noqa: F401
+    PLATFORM_SCHEMA,
+    PLATFORM_SCHEMA_BASE,
+)
+from homeassistant.helpers.entity import ToggleEntity
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+from homeassistant.loader import bind_hass
+
+from .const import (
+    ATTR_AVAILABLE_TONES,
+    ATTR_DEFAULT_DURATION,
+    ATTR_DEFAULT_TONE,
+    ATTR_DURATION,
+    ATTR_TONE,
+    ATTR_VOLUME_LEVEL,
+    DOMAIN,
+    SERVICE_SET_DEFAULT_DURATION,
+    SERVICE_SET_DEFAULT_TONE,
+    SERVICE_SET_VOLUME_LEVEL,
+    SUPPORT_DURATION,
+    SUPPORT_TONES,
+    SUPPORT_TURN_OFF,
+    SUPPORT_TURN_ON,
+    SUPPORT_VOLUME_SET,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=60)
+
+TURN_ON_SCHEMA = {
+    vol.Optional(ATTR_TONE): cv.string,
+    vol.Optional(ATTR_DURATION): cv.positive_float,
+    vol.Optional(ATTR_VOLUME_LEVEL): cv.small_float,
+}
+
+
+@bind_hass
+def is_on(hass: HomeAssistantType, entity_id: str) -> bool:
+    """
+    Return if the siren is on based on the state machine.
+
+    Async friendly.
+    """
+    return hass.states.is_state(entity_id, STATE_ON)
+
+
+def filter_turn_on_params(siren: SirenEntity, params: dict[str, Any]):
+    """Filter out params not supported by the siren."""
+    supported_features = siren.supported_features or 0
+
+    if not supported_features & SUPPORT_TONES:
+        params.pop(ATTR_TONE, None)
+    if not supported_features & SUPPORT_DURATION:
+        params.pop(ATTR_DURATION, None)
+    if not supported_features & SUPPORT_VOLUME_SET:
+        params.pop(ATTR_VOLUME_LEVEL, None)
+
+    return params
+
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
+    """Set up siren devices."""
+    component = hass.data[DOMAIN] = EntityComponent(
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+
+    async def async_handle_turn_on_service(
+        siren: SirenEntity, call: ServiceCall
+    ) -> None:
+        """Handle turning a siren on."""
+        await siren.async_turn_on(
+            **filter_turn_on_params(siren, dict(call.data.get("params", {})))
```

Comment:
`"params"` isn't part of the service call data by default. That's something that the light integration has added.

Reply:
Thanks, added a test for this so it doesn't get missed